### PR TITLE
feat: Search by metadata and filter by label

### PR DESCRIPTION
**Pull request by SilviaZeta**
# Description

Adds: 

- A search box that allows to search for images by any metadata field (one autocomplete combobox for key and for value);
- An "Image labels" accordion that lists all unique (image-wise) labels and allows to filter for images by label.

addresses #5 (search only)

# Dependency changes

Adds @material-ui/lab

# Testing

No

# Documentation

N/A

# Migrations (if applicable)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] My changes generate no new warnings
- [ ] ~New migrations have been committed~
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [ ] ~New and existing unit tests pass locally with my changes~

### CODE REVIEW

**Comment by ChasNelson1990** [comment]
Do we have an examples page that loads some data for azure to use? I followed the link on my phone but its empty of tiles.

File: package.json [hunk 1]
```diff
@@ -60,6 +60,8 @@
     "react-dom": "^17.0.1"
   },
   "dependencies": {
-    "@material-ui/core": "^4.11.3"
+    "@material-ui/core": "^4.11.3",
+    "@material-ui/icons": "^4.11.2",
+    "@material-ui/lab": "*"
```
Comment (cooper667) [inline package.json:65]: can we use an actual version here?

File: src/search/ComboBox.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  const [inputKey, setInputKey] = useState("");
+  const [inputOptions, setOptions] = useState([]);
+  const [inputValue, setInputValue] = useState("");
+
+  const updateOptions = (): void => {
+    const options: Set<string> = new Set();
+    metadata.forEach((mitem: MetaItem) => {
+      for (const [key, value] of Object.entries(mitem)) {
+        if (key === inputKey) {
+          if (typeof value === "object") {
```
Comment (cooper667) [inline src/search/ComboBox.tsx:53]: Should this be `if (Array.isArray(value))`?

File: src/search/interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+export type Metadata = Array<MetaItem>;
+
+export type MetaItem = {
```
Comment (cooper667) [inline src/search/interfaces.ts:3]:
```suggestion
export type MetaItem = {
  [index: string]: string | string[];
```

Comment (cooper667) [reply src/search/interfaces.ts:3]: This should let you access this in the normal way?

File: src/ui.tsx [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      .split(".")
+      .shift();
+    return this.state.imageNames.includes(tileName);
+  };
+
+  loadMeta = (
+    jsonUrl: string,
+    onLoadCallback: (data: Metadata) => void
+  ): void => {
+    // const myRequest = new Request(jsonUrl);
```
Comment (cooper667) [inline src/ui.tsx:131]:
```suggestion
```

**Review by cooper667 — APPROVED** [review]
Works for me

File: src/search/LabelsAccordion.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+              dense
+              button
+              onClick={addDelLabel(label)}
+              className={style.listItem}
+            >
+              {(() => {
+                if (selectedLabels.includes(label))
+                  return <Label className={style.icon} />;
+                return <LabelOutlined className={style.icon} />;
+              })()}
```
Comment (philipjackson) [inline src/search/LabelsAccordion.tsx:87]: {selectedLabels.includes(label) ? (
                <Label className={style.icon} />
              ) : (
                <LabelOutlined className={style.icon} />
              )}

I tried this, it's a little nicer and it seems to work still

File: src/search/ComboBox.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  metadataKeys,
+  callback,
+}: Props): ReactElement {
+  const style = useStyles();
+  const [inputKey, setInputKey] = useState("");
+  const [inputOptions, setOptions] = useState([]);
+  const [inputValue, setInputValue] = useState("");
+
+  const updateOptions = (): void => {
+    if (metadataKeys.indexOf(inputKey) === -1) return;
```
Comment (cooper667) [inline src/search/ComboBox.tsx:49]:
```suggestion
    if (!metadataKeys.includes(inputKey)) return;
```

File: src/search/interfaces.ts (hunk 3 above)
Comment (SilviaZeta) [reply src/search/interfaces.ts:3]: This is the fix I was looking for! Thanks!

**Review by cooper667 — APPROVED** [review]

**Comment by philipjackson** [comment]
If we do `const [selectedLabels, setSelectedLabels] = useState<string[]>(imageLabels);` on line 49 in `LabelsAccordion.tsx` and
```
        {this.state.imageLabels.length > 0 && (
          <Grid item style={{ position: "relative", width: "200px" }}>
            <LabelsAccordion
              expanded={this.state.expanded === "labels-toolbox"}
              handleToolboxChange={this.handleToolboxChange("labels-toolbox")}
              imageLabels={this.state.imageLabels}
              callback={this.handleOnLabelSelection}
            />
          </Grid>
        )}
```
on line 156 in `ui.tsx`, then we can prevent the labels accordion getting desynced with the image tiles - currently when the page loads, the labels are all deselected but it shows all the tiles anyway, so it should start with all the labels selected.

**Comment by SilviaZeta** [comment]
> If we do `const [selectedLabels, setSelectedLabels] = useState<string[]>(imageLabels);` on line 49 in `LabelsAccordion.tsx` and
> 
> ```
>         {this.state.imageLabels.length > 0 && (
>           <Grid item style={{ position: "relative", width: "200px" }}>
>             <LabelsAccordion
>               expanded={this.state.expanded === "labels-toolbox"}
>               handleToolboxChange={this.handleToolboxChange("labels-toolbox")}
>               imageLabels={this.state.imageLabels}
>               callback={this.handleOnLabelSelection}
>             />
>           </Grid>
>         )}
> ```
> 
> on line 156 in `ui.tsx`, then we can prevent the labels accordion getting desynced with the image tiles - currently when the page loads, the labels are all deselected but it shows all the tiles anyway, so it should start with all the labels selected. It's difficult cause LabelsAccordion first renders before the metadata is actually loaded, so if we try to initialize `selectedLabels` as `imageLabels` then it'll be empty. Ideally we could have LabelsAccordion detect when the `imageLabels` prop changes, but I'm not sure there's an easy way to do that with function components. In the meantime we can just make sure that it doesn't render until the metadata is loaded, and that seems to work.

Thanks, well spotted! Instead of `this.state.imageLabels.length > 0` i've added a `useEffect` that listens for changes in `imageLabels` and updates `selectedLabels` accordingly (as you suggested :-)
```
  useEffect(() => {
    setSelectedLabels(imageLabels);
  }, [imageLabels]);
```

**Review by cooper667 — APPROVED** [review]

**Review by ChrisBaidoo — APPROVED** [review]

**Review by philipjackson — APPROVED** [review]

File: src/ui.tsx [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        if (key === inputKey) {
+          if (
+            (typeof value === "object" && value.includes(inputValue)) ||
+            value === inputValue
+          ) {
+            filteredMeta.push(mitem);
+            break;
+          }
+        }
+      }
```
Comment (philipjackson) [inline src/ui.tsx:72]: Could this be done without the inner for loop after Craig's typescript fix?

Comment (SilviaZeta) [reply src/ui.tsx:72]: Yes, i've changed that in commit `fix: MetaItem is search/interfaces.ts (part2)`.

**Pull request merged**